### PR TITLE
Add Retry To Lambda Functions and Add the Error to the DB

### LINKED ISSUE #6349: Noobaa Functions - Logging of failed invocations and retry them
When a function call fails within the Nooba Endpoint, there is no mechanism available to retry the function-call.

It would be awsom if there would be an integration of a retry mechanism. 

We thought about a logging mechnism in the first place to get all failed invocations from noobaa with some info of the event causing the fail. 
If possible it would be great to retry the function call a second time or n-times if it can be configured by the user.

### PR DESCRIPTION
**Pull request by liranmauda**
### Explain the changes
Add retry to lambda functions:
- Added retries (default is 2)
- Added UI Field in the add/edit retries
- async/await some flows

Signed-off-by: liranmauda <liran.mauda@gmail.com>

### Issues: 
~1. Edit form is not loading the current number of attempts~
~2. There is no tooltip that explains the attempts field~ 

### out of scope
~1. Writing a failed invoke log into the DB~
2. converting the lambda function and save it in the DB

### Testing Instructions:
For the retries: 
1. Add/Edit a trigger With 1 or more retires
2. See that on "fail" it retries
3. See that on "success" It does not retry
4. Check that if 0 retries are set, then the trigger runs once.

For the failed error in the db:
1. run a failed function
2. see in the db that the `error_msg` is there:
    run `db.func_stats.find({'error':true}).pretty()`
```
{
	"_id" : ObjectId("601ab5ac0d4f4c000d395964"),
	"system" : ObjectId("60196a75933e03002de1f296"),
	"func" : ObjectId("60196c60933e03002de1f2b7"),
	"time" : ISODate("2021-02-03T14:39:18.870Z"),
	"took" : 21597,
	"error" : true,
	"error_msg" : "Func handler not a function main.no"
}
```

### CODE REVIEW

**Comment by liranmauda** [comment]
@QuintinBecker @lallinger-arbeit

**Comment by lallinger-arbeit** [comment]
@liranmauda Thank you very much for such a fast implementation! If it is in a state where we should test it just hit me up!

**Comment by liranmauda** [comment]
> @liranmauda Thank you very much for such a fast implementation! If it is in a state where we should test it just hit me up!

@lallinger-arbeit It is in a state to test it.
I will later add the "out of scope"  parts (maybe in another PR).

**Comment by lallinger-arbeit** [comment]
> 
> 
> > @liranmauda Thank you very much for such a fast implementation! If it is in a state where we should test it just hit me up!
> 
> @lallinger-arbeit It is in a state to test it.
> I will later add the "out of scope" parts (maybe in another PR).

ok, then we will try to test it this week!

**Comment by liranmauda** [comment]
Added the failed invoke error message into the func_stats collection.

**Comment by lallinger-arbeit** [comment]
@liranmauda looks very good so far. I just noticed that in the UI the attempts field is always displayed empty, but it should show the configured attempts.
Also as the information of failed functions is saved in the `func_stats` collection, it would be best the problem with the [slow mapreduce job of the MongoDB](https://github.com/noobaa/noobaa-core/issues/6265) on this collection would be fixed too. Because currently we disabled the code which writes to the `func_stats`, but would definitely need it for this retry feature

**Comment by lallinger-arbeit** [comment]
@liranmauda it would also be nice to have a message in `func_stats` which states if a event exceeded its maximum retries

**Comment by liranmauda** [comment]
> @liranmauda looks very good so far. I just noticed that in the UI the attempts field is always displayed empty, but it should show the configured attempts.

Yes, I am aware of this issue, waiting for @nb-ohad Input.

**Comment by liranmauda** [comment]
Added tooltip.
Not sure it is very helpful, any suggestion for the text will be welcome.

**Comment by liranmauda** [comment]
> > @liranmauda looks very good so far. I just noticed that in the UI the attempts field is always displayed empty, but it should show the configured attempts.
> 
> Yes, I am aware of this issue, waiting for @nb-ohad Input.

Fixed with @nb-ohad help.

**Comment by lallinger-arbeit** [comment]
@liranmauda The attempts field now shows up as desired. Thinking about it i think it would be more straightforward if the field would be called `Retries` instead of `Attempts`. As attempts=0 would mean the trigger shouldn't be run...

If you rename it to retries i would suggest the following tooltip: "Number of retries Noobaa will make if the trigger fails"

**Comment by liranmauda** [comment]
> @liranmauda The attempts field now shows up as desired. Thinking about it i think it would be more straightforward if the field would be called `Retries` instead of `Attempts`. As attempts=0 would mean the trigger shouldn't be run...
> 
> If you rename it to retries i would suggest the following tooltip: "Number of retries Noobaa will make if the trigger fails"

We can change it to retries. 
Currently, If you try to put 0 in the attempts field it will not let you.
If we change it to retries we should allow 0

**Comment by lallinger-arbeit** [comment]
> We can change it to retries.
> Currently, If you try to put 0 in the attempts field it will not let you.
> If we change it to retries we should allow 0

I would prefere this, sounds better than attempts

**Comment by liranmauda** [comment]
Changed to retries in the UI, changed the tooltip as suggested, and allowed 0 to be an input (as now it is "retry" and not "attempt")

**Comment by lallinger-arbeit** [comment]
@liranmauda update me when you have added the retry field to the log message in the db, i will retest then

**Comment by liranmauda** [comment]
> > Changed to retries in the UI, changed the tooltip as suggested, and allowed 0 to be an input (as now it is "retry" and not "attempt")
> 
> thanks!
> 
> @liranmauda update me when you have added the retry field to the log message in the db, i will retest then

@lallinger-arbeit I think we should open a gap and do this in a different PR, And merge this one.
The `P.retry` that we are using is not reflecting the current number of attempts left.

**Comment by lallinger-arbeit** [comment]
> @lallinger-arbeit I think we should open a gap and do this in a different PR, And merge this one.
> The `P.retry` that we are using is not reflecting the current number of attempts left.

i'm fine with that

File: src/server/system_services/schemas/bucket_schema.js [hunk 1]
```diff
@@ -319,6 +319,9 @@ module.exports = {
                     object_suffix: {
                         type: 'string'
                     },
+                    attempts: {
```
Comment (lallinger-arbeit) [inline src/server/system_services/schemas/bucket_schema.js:322]: @liranmauda i think the field should be called `retries` in the API too

Comment (liranmauda) [reply src/server/system_services/schemas/bucket_schema.js:322]: The API is with Attempts and not retries as it reflects the behavior better.
We are using `P.retry` and it uses attempts. 
Providing 1 will run it once without retries, providing more will do retries.
https://github.com/noobaa/noobaa-core/blob/de261f32fb59a973fb244e84512f572be0a892f4/src/util/promise.js#L191-L195
I have changed the UI to be more straightforward, The BE left with attempts.

Comment (lallinger-arbeit) [reply src/server/system_services/schemas/bucket_schema.js:322]: ok

**Review by jeniawhite — APPROVED** [review]

**Pull request merged**